python/cudf/bindings/copying.pyx

```diff
@@ -0,0 +1,236 @@
+    cdef gdf_column **c_cols = <gdf_column**>malloc(sizeof(gdf_column*)*col_count)
+
+    cdef i
+    for i in range(col_count):
+        check_gdf_compatibility(cols[i])
+        c_cols[i] = column_view_from_column(cols[i])
+
+    return c_cols
+
+
+cdef free_table(cudf_table* table, gdf_column** cols):
+    cdef i
+    cdef gdf_column *c_col
+    for i in range(table[0].num_columns()) :
+        c_col = table[0].get_column(i)
+        free(c_col)
+
+    del table
+    free(cols)
+
+
+def clone_columns_with_size(in_cols, row_size):
+    from cudf.dataframe import columnops
+    out_cols = []
+    for col in in_cols:
+        o_col = columnops.column_empty(row_size,
+                                       dtype = col.dtype,
+                                       masked = col.has_null_mask)
+        out_cols.append(o_col)
+
+    return out_cols
+
+
+def apply_gather(in_cols, maps, out_cols=None):
+    """
+      Call cudf::gather.
+
+     * in_cols input column array
+     * maps RMM device array with gdf_index_type (np.int32 compatible dtype)
+     * out_cols the destination column array to output
+
+     * returns out_cols
+    """
+    maps = astype(maps, 'int32')
+
+    col_count=len(in_cols)
+    gather_count = len(maps)
+    cdef gdf_column** c_in_cols = cols_view_from_cols(in_cols)
+    cdef cudf_table* c_in_table = new cudf_table(c_in_cols, col_count)
+
+    # check out_cols == in_cols and out_cols=None cases
+    cdef bool is_same_input = False
+    cdef gdf_column** c_out_cols
+    cdef cudf_table* c_out_table
+    if out_cols == in_cols :
+        is_same_input = True
+        c_out_table = c_in_table
+    elif out_cols != None :
+        c_out_cols = cols_view_from_cols(out_cols)
+        c_out_table = new cudf_table(c_out_cols, col_count)
+    else:
+        out_cols = clone_columns_with_size(in_cols, gather_count)
+        c_out_cols = cols_view_from_cols(out_cols)
+        c_out_table = new cudf_table(c_out_cols, col_count)
+
+    cdef uintptr_t c_maps_ptr
+    cdef gdf_index_type* c_maps
+    if gather_count != 0 :
+        # size check, cudf::gather requires same length for maps and out table.
+        assert gather_count == out_cols[0].data.size
+
+        c_maps_ptr = get_ctype_ptr(maps)
+        c_maps = <gdf_index_type*>c_maps_ptr
+
+        with nogil:
+            gather(c_in_table, c_maps, c_out_table)
+
+    if is_same_input == False :
+        free_table(c_out_table, c_out_cols)
+    
+    free_table(c_in_table, c_in_cols)
+    
+    return out_cols
+
+
+def apply_gather_column(in_col, maps, out_col=None):
+    """
+      Call cudf::gather.
+
+     * in_cols input column
+     * maps device array
+     * out_cols the destination column to output
+
+     * returns out_col
+    """
+
+    in_cols = [in_col]
+    out_cols = None if out_col == None else [out_col]
+
+    out_cols = apply_gather(in_cols, maps, out_cols)
+
+    return out_cols[0]
+
+
+def apply_gather_array(dev_array, maps, out_col=None):
```

Comment:
Should we merge this function into `apply_gather_column` above so that it can accept either a device array or a column?  Ditto with `apply_scatter_column` and `apply_scatter_array` below.

Reply:
Hi, @shwina 
I'm afraid I'm c++ guy and  not familiar to python. How can I merge the functions to accept either of a device array or a column?

Reply:
We could just create a single function `apply_gather` (I would suggest renaming the current `apply_gather` function above to something like `cpp_apply_gather`):

```python

def apply_gather(in, maps, out):
    in_col = columnops.as_column(in) # no-op if `in` is a column already
    out_cols = None if out == None else [out]
    out_cols = apply_gather(in_cols, maps, out_cols)
    return out_cols[0]
```

Reply:
I think we can hold off and do this in a future refactor. There's likely a bigger problem that there's places we should be using columns instead of arrays and/or arrays instead of columns.